ruma-events-macros/src/event_enum.rs

```diff
@@ -109,26 +115,28 @@ pub fn expand_event_enum(input: EventEnumInput) -> syn::Result<TokenStream> {
 }
 
 fn expand_any_enum_with_deserialize(
-    input: &EventEnumInput,
-    var: EventKindVariation,
+    name: &EventKind,
+    events: &[LitStr],
+    attrs: &[Attribute],
+    var: &EventKindVariation,
 ) -> syn::Result<TokenStream> {
-    let event_struct = if let Some(i) = input.name.to_event_ident(&var) {
+    let event_struct = if let Some(i) = name.to_event_ident(var) {
         i
     } else {
         return Ok(TokenStream::new());
     };
-    let ident = if let Some(i) = input.name.to_event_enum_ident(&var) {
+    let ident = if let Some(i) = name.to_event_enum_ident(var) {
         i
     } else {
         return Ok(TokenStream::new());
     };
 
-    let attrs = &input.attrs;
-    let event_type_str = &input.events;
+    let attrs = attrs;
+    let event_type_str = events;
 
-    let variants = input.events.iter().map(to_camel_case).collect::<syn::Result<Vec<_>>>()?;
+    let variants = events.iter().map(to_camel_case).collect::<syn::Result<Vec<_>>>()?;
```

Comment:
If you move this out of the function (maybe it's also used elsewhere?) you can change it to return `Option<TokenStream>` and use `name.to_event_ident(var)?` and `name.to_event_enum_ident(var)?` above :)

By the way, how often are those used together? Since it's guaranteed either none or both fail, maybe a combined `Option<(Ident, Ident)>` returning function would be useful? (this could be an alternative way to make the early return code above more concise.

Reply:
This is the only time they are used. Is there a way to map an `Err(TokenStream)` to an `Ok(TokenStream)` then this would be really nice.

I'm assuming the code window is just off and you are talking about putting these together

```rust
let event_struct = if let Some(i) = name.to_event_ident(var) {
        i
    } else {
        return Ok(TokenStream::new());
    };
    let ident = if let Some(i) = name.to_event_enum_ident(var) {
        i
    } else {
        return Ok(TokenStream::new());
    };
```

Reply:
The code window isn't off, the first part of my comment was referring exactly to the line shown with "this". You're right that the latter part of the comment is about the code you just showed though.

Reply:
`to_camel_case` returns a useful error would we lose that or am I not following what your idea is very well?

Reply:
I was saying you could move the construction of `variants` to outside this function (it seems to also be used elsewhere, maybe this even makes it so it doesn't have to be created twice) and pass it in as an argument. Then this function doesn't have a way to fail anymore and doesn't need to return `syn::Result`, so it can use `?` for the `Option` stuff instead (`Option<TokenStream>` can be interpolated into a `quote!` invocation just the same as `TokenStream`, producing no output on `None`).

Reply:
If this doesn't make things less readable / obvious though, feel free to ignore this suggestion and resolve the thread (but maybe reduce the two early returns to one with `let (event_struct, ident) = match (name.to_event_ident(var), name.to_event_enum_ident(var)) { ... };`).